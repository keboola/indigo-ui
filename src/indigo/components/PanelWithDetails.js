import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { Panel } from 'react-bootstrap';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faAngleDown } from '@fortawesome/free-solid-svg-icons/faAngleDown';
import { faAngleRight } from '@fortawesome/free-solid-svg-icons/faAngleRight';
import { faAngleUp } from '@fortawesome/free-solid-svg-icons/faAngleUp';

const PLACEMENT_BOTTOM = 'bottom';
const PLACEMENT_TOP = 'top';

class PanelWithDetails extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      expanded: props.defaultExpanded,
      panelHeaderTitle: props.defaultExpanded ? props.labelCollapse : props.labelOpen,
    };

    this.toggleExpanded = this.toggleExpanded.bind(this);
  }

  render() {
    return (
      <Panel
        onToggle={this.toggleExpanded}
        expanded={this.state.expanded}
        className={classnames('panel-show-details', `panel-show-details-${this.props.placement}`, {
          'panel-preview': !!this.props.preview,
          [`panel-preview-${this.props.preview}`]: !!this.props.preview,
        })}
      >
        <Panel.Heading>
          <Panel.Title toggle>
            <FontAwesomeIcon fixedWidth icon={this.getIcon()} className="icon-addon-right" />
            {this.state.panelHeaderTitle}
          </Panel.Title>
        </Panel.Heading>
        <Panel.Body collapsible>{this.props.children}</Panel.Body>
      </Panel>
    );
  }

  getIcon() {
    if (this.props.placement === PLACEMENT_BOTTOM) {
      return this.state.expanded ? faAngleUp : faAngleRight;
    }
    return this.state.expanded ? faAngleDown : faAngleRight;
  }

  toggleExpanded() {
    this.setState((prevState) => ({
      expanded: !prevState.expanded,
      panelHeaderTitle: !prevState.expanded ? this.props.labelCollapse : this.props.labelOpen,
    }));
  }
}

PanelWithDetails.propTypes = {
  defaultExpanded: PropTypes.bool,
  labelCollapse: PropTypes.node,
  labelOpen: PropTypes.node,
  children: PropTypes.any.isRequired,
  placement: PropTypes.oneOf([PLACEMENT_BOTTOM, PLACEMENT_TOP]),
  preview: PropTypes.oneOf(['normal', 'small']),
};

PanelWithDetails.defaultProps = {
  defaultExpanded: false,
  labelCollapse: 'Hide details',
  labelOpen: 'Show details',
  placement: PLACEMENT_TOP,
};

export default PanelWithDetails;
