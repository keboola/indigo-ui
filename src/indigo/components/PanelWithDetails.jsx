import React, {PropTypes} from 'react';
import {Panel} from 'react-bootstrap';

export default React.createClass({

  propTypes: {
    defaultExpanded: PropTypes.bool,
    labelCollapsed: PropTypes.string,
    labelOpen: PropTypes.string,
    children: PropTypes.any
  },

  getDefaultProps() {
    return ({
      defaultExpanded: false,
      labelCollapsed: 'Show details',
      labelOpen: 'Hide details'
    });
  },

  getInitialState() {
    return ({
      panelHeaderTitle: (this.props.defaultExpanded) ? this.props.labelOpen : this.props.labelCollapsed
    });
  },

  render() {
    return (
      <Panel
        onEnter={() => this.setState({panelHeaderTitle: this.props.labelOpen})}
        onExit={() => this.setState({panelHeaderTitle: this.props.labelCollapsed})}
        header={this.state.panelHeaderTitle}
        defaultExpanded={this.props.defaultExpanded}
        className="panel-show-details"
        collapsible={true}>
          {this.props.children}
      </Panel>
    );
  }
});
