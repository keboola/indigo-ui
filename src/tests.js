import renderer from 'react-test-renderer';

export const snapshot = (component) => {
  const tree = renderer
    .create(component)
    .toJSON();
  expect(tree).toMatchSnapshot();
};
