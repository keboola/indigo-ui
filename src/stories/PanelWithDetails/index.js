import React from 'react';
import { storiesOf } from '@storybook/react';
import PanelWithDetails from '../../indigo/components/PanelWithDetails';
import { withInfo } from '@storybook/addon-info';


storiesOf('PanelWithDetails', module)
    .add(
        'Basic init',
        withInfo({
            text: `PanelWithDetails serves to reveal more details appended to main content`,
            inline: true,
        })(() => {
            return (
                <PanelWithDetails>
                    Lorem ipsum dolor sit amet, consectetur adipisicing elit.
                </PanelWithDetails>
            )
        })
    )
    .add(
        'Initially expanded',
        withInfo({
            text: `
        Allows to set intial expanded/collapsed state.
        `,
            inline: true,
        })(() => {
            return (
                <PanelWithDetails
                    defaultExpanded={true}>
                    Lorem ipsum dolor sit amet, consectetur adipisicing elit.
                </PanelWithDetails>
            )
        })
    )
    .add(
        'Custom labels',
        withInfo({
            text: `
        Allows to change labels for both expanded/collapsed state
        `,
            inline: true,
        })(() => {
            return (
                <PanelWithDetails
                    labelCollapse="Less"
                    labelOpen="More"
                >
                    Lorem ipsum dolor sit amet, consectetur adipisicing elit.
                </PanelWithDetails>
            )
        })
    )
    .add(
        'Bottom controls',
        withInfo({
            text: `
        COntrol label forces to bottom
        `,
            inline: true,
        })(() => {
            return (
                <PanelWithDetails
                    placement="bottom"
                >
                    Lorem ipsum dolor sit amet, consectetur adipisicing elit.
                </PanelWithDetails>
            )
        })
    )
  .add(
    'List + Bottom controls',
    withInfo({
      text: `
          Example how to hide part of unordered list + bottom controls.
          `,
      inline: true,
    })(() => {
      return (
        <div style={{backgroundColor: '#fff', maxWidth: '240px'}}>
          <h4>Input</h4>
          <ul className="list-no-bottom-margin">
            <li><span className="kbc-sapi-table-link">in.c-csv-import.100</span></li>
            <li><span className="kbc-sapi-table-link">in.c-csv-import.101</span></li>
            <li><span className="kbc-sapi-table-link">in.c-csv-import.102</span></li>
            <li><span className="kbc-sapi-table-link">in.c-csv-import.103</span></li>
            <li><span className="kbc-sapi-table-link">in.c-csv-import.104</span></li>
          </ul>
          <PanelWithDetails
            placement="bottom"
            labelOpen="Show 5 more tables"
            labelCollapse="Show less"
          >
            <ul>
              <li><span className="kbc-sapi-table-link">in.c-csv-import.105</span></li>
              <li><span className="kbc-sapi-table-link">in.c-csv-import.106</span></li>
              <li><span className="kbc-sapi-table-link">in.c-csv-import.107</span></li>
              <li><span className="kbc-sapi-table-link">in.c-csv-import.108</span></li>
              <li><span className="kbc-sapi-table-link">in.c-csv-import.109</span></li>
            </ul>
          </PanelWithDetails>
        </div>
      )
    })
  );
