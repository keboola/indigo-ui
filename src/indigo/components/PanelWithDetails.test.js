import React from 'react';

import { snapshot } from "../../tests";
import PanelWithDetails from './PanelWithDetails';

describe('<PanelWithDetails />', () => {

  it('PanelWithDetails - default', () => {
    snapshot(
      <PanelWithDetails>
        content
      </PanelWithDetails>
    );
  });

  it('PanelWithDetails - default expanded', () => {
    snapshot(
      <PanelWithDetails defaultExpanded={true}>
        content
      </PanelWithDetails>
    );
  });

  it('PanelWithDetails - custom labels', () => {
    snapshot(
      <PanelWithDetails labelOpen="Show" labelCollapse="Hide">
        content
      </PanelWithDetails>
    );
  });

  it('Bottom controls', () => {
    snapshot(
      <PanelWithDetails placement="bottom">
        content
      </PanelWithDetails>
    );
  });

  it('PanelWithDetails - small preview', () => {
    snapshot(
      <PanelWithDetails preview="small">
        content
      </PanelWithDetails>
    );
  });

});
