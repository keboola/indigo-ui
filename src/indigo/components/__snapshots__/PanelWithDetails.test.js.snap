// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PanelWithDetails /> PanelWithDetails - bottom controls 1`] = `
<div
  className="panel-show-details panel-show-details-bottom panel panel-default"
>
  <div
    className="panel-heading"
  >
    <div
      className="panel-title"
    >
      <a
        aria-expanded={false}
        className="collapsed"
        href="#"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="button"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-angle-right fa-fw icon-addon-right"
          data-icon="angle-right"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 256 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M64 448c-8.188 0-16.38-3.125-22.62-9.375c-12.5-12.5-12.5-32.75 0-45.25L178.8 256L41.38 118.6c-12.5-12.5-12.5-32.75 0-45.25s32.75-12.5 45.25 0l160 160c12.5 12.5 12.5 32.75 0 45.25l-160 160C80.38 444.9 72.19 448 64 448z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
        Show details
      </a>
    </div>
  </div>
  <div
    aria-expanded={null}
    className="panel-collapse collapse"
  >
    <div
      className="panel-body"
    >
      content
    </div>
  </div>
</div>
`;

exports[`<PanelWithDetails /> PanelWithDetails - custom labels 1`] = `
<div
  className="panel-show-details panel-show-details-top panel panel-default"
>
  <div
    className="panel-heading"
  >
    <div
      className="panel-title"
    >
      <a
        aria-expanded={false}
        className="collapsed"
        href="#"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="button"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-angle-right fa-fw icon-addon-right"
          data-icon="angle-right"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 256 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M64 448c-8.188 0-16.38-3.125-22.62-9.375c-12.5-12.5-12.5-32.75 0-45.25L178.8 256L41.38 118.6c-12.5-12.5-12.5-32.75 0-45.25s32.75-12.5 45.25 0l160 160c12.5 12.5 12.5 32.75 0 45.25l-160 160C80.38 444.9 72.19 448 64 448z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
        Show
      </a>
    </div>
  </div>
  <div
    aria-expanded={null}
    className="panel-collapse collapse"
  >
    <div
      className="panel-body"
    >
      content
    </div>
  </div>
</div>
`;

exports[`<PanelWithDetails /> PanelWithDetails - default 1`] = `
<div
  className="panel-show-details panel-show-details-top panel panel-default"
>
  <div
    className="panel-heading"
  >
    <div
      className="panel-title"
    >
      <a
        aria-expanded={false}
        className="collapsed"
        href="#"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="button"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-angle-right fa-fw icon-addon-right"
          data-icon="angle-right"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 256 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M64 448c-8.188 0-16.38-3.125-22.62-9.375c-12.5-12.5-12.5-32.75 0-45.25L178.8 256L41.38 118.6c-12.5-12.5-12.5-32.75 0-45.25s32.75-12.5 45.25 0l160 160c12.5 12.5 12.5 32.75 0 45.25l-160 160C80.38 444.9 72.19 448 64 448z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
        Show details
      </a>
    </div>
  </div>
  <div
    aria-expanded={null}
    className="panel-collapse collapse"
  >
    <div
      className="panel-body"
    >
      content
    </div>
  </div>
</div>
`;

exports[`<PanelWithDetails /> PanelWithDetails - default expanded 1`] = `
<div
  className="panel-show-details panel-show-details-top panel panel-default"
>
  <div
    className="panel-heading"
  >
    <div
      className="panel-title"
    >
      <a
        aria-expanded={true}
        className=""
        href="#"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="button"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-angle-down fa-fw icon-addon-right"
          data-icon="angle-down"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 384 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M192 384c-8.188 0-16.38-3.125-22.62-9.375l-160-160c-12.5-12.5-12.5-32.75 0-45.25s32.75-12.5 45.25 0L192 306.8l137.4-137.4c12.5-12.5 32.75-12.5 45.25 0s12.5 32.75 0 45.25l-160 160C208.4 380.9 200.2 384 192 384z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
        Hide details
      </a>
    </div>
  </div>
  <div
    aria-expanded={null}
    className="panel-collapse collapse in"
  >
    <div
      className="panel-body"
    >
      content
    </div>
  </div>
</div>
`;

exports[`<PanelWithDetails /> PanelWithDetails - small preview 1`] = `
<div
  className="panel-show-details panel-show-details-top panel-preview panel-preview-small panel panel-default"
>
  <div
    className="panel-heading"
  >
    <div
      className="panel-title"
    >
      <a
        aria-expanded={false}
        className="collapsed"
        href="#"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="button"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-angle-right fa-fw icon-addon-right"
          data-icon="angle-right"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 256 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M64 448c-8.188 0-16.38-3.125-22.62-9.375c-12.5-12.5-12.5-32.75 0-45.25L178.8 256L41.38 118.6c-12.5-12.5-12.5-32.75 0-45.25s32.75-12.5 45.25 0l160 160c12.5 12.5 12.5 32.75 0 45.25l-160 160C80.38 444.9 72.19 448 64 448z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
        Show details
      </a>
    </div>
  </div>
  <div
    aria-expanded={null}
    className="panel-collapse collapse"
  >
    <div
      className="panel-body"
    >
      content
    </div>
  </div>
</div>
`;
